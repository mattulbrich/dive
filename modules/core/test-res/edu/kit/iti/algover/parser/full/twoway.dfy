// GOAL: Know how to parse
//   https://github.com/Microsoft/dafny/blob/master/Test/vstte2012/Two-Way-Sort.dfy


// method swap<T>(a: array<T>, i: int, j: int)
method swap(a: array<bool>, i: int, j: int)
  requires a != null
  // requires 0 <= i < j < a.Length
  requires 0 <= i && i < j && j < a.Length
  ensures a[i] == old(a[j])
  ensures a[j] == old(a[i])
  // ensures forall m :: 0 <= m < a.Length && m != i && m != j ==> a[m] == old(a[m]);
  ensures (forall m:int :: 0 <= m && m < a.Length && m != i && m != j ==> a[m] == old(a[m]))
  // ensures multiset(a[..]) == old(multiset(a[..]));
  modifies a
{
  var t:bool := a[i];
  a[i] := a[j];
  a[j] := t;
}

method two_way_sort(a: array<bool>)
  requires a != null
  // ensures forall m,n :: 0 <= m < n < a.Length ==> (!a[m] || a[n])
  ensures (forall m:int :: (forall n:int :: 0 <= m && m < n && n < a.Length ==> (!a[m] || a[n])))
  // ensures multiset(a[..]) == old(multiset(a[..]));
  modifies a
{
  var i:int := 0;
  var j:int := a.Length - 1;
  while (i <= j)
    // invariant 0 <= i <= j + 1 <= a.Length;
    invariant 0 <= i && i <= j + 1 && j+1 <= a.Length
    invariant (forall m:int :: 0 <= m && m < i ==> !a[m])
    invariant (forall n:int :: j < n && n < a.Length ==> a[n])
    // invariant multiset(a[..]) == old(multiset(a[..]))
  {
    if (!a[i]) {
      i := i+1;
    } else { if (a[j]) {
      j := j-1;
    } else {
      swap(a, i, j);
      i := i+1;
      j := j-1;
    }}
  }
}