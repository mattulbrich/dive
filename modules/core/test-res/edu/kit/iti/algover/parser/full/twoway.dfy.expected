
(COMPILATION_UNIT 
 (method swap 
  (ARGS 
   (VAR a 
    (array bool)) 
   (VAR i int) 
   (VAR j int)) 
  (requires 
   (!= a null)) 
  (requires 
   (&& 
    (&& 
    (<= 0 i) 
    (< i j))
   (< j (Length a)))) 
  (ensures 
   (== 
    (ARRAY_ACCESS a i) 
    (old 
      (ARRAY_ACCESS a j))))
  (ensures 
   (== 
    (ARRAY_ACCESS a j) 
    (old 
      (ARRAY_ACCESS a i))))
  (ensures 
   (forall m int 
    (==> 
     (&& 
      (&& 
       (&& 
        (<= 0 m) 
        (< m (Length a)))
       (!= m i))
      (!= m j)) 
     (== 
      (ARRAY_ACCESS a m) 
      (old 
        (ARRAY_ACCESS a m))))))
  (BLOCK 
   (var t bool 
    (ARRAY_ACCESS a i)) 
   (:= 
    (ARRAY_ACCESS a i) 
    (ARRAY_ACCESS a j)) 
   (:= 
    (ARRAY_ACCESS a j) t))) 
 (method two_way_sort 
  (ARGS 
   (VAR a 
    (array bool))) 
  (requires 
   (!= a null)) 
  (ensures 
   (forall m int 
    (forall n int 
     (==> 
      (&& 
       (&& 
        (<= 0 m) 
        (< m n))
        (< n (Length a))) 
      (|| 
       (! 
        (ARRAY_ACCESS a m)) 
       (ARRAY_ACCESS a n)))))) 
  (BLOCK 
   (var i int 0) 
   (var j int 
    (- 
     (Length a) 1)) 
   (while 
    (<= i j) 
    (invariant 
     (&& 
      (&& 
       (<= 0 i) 
       (<= i (+ j 1))) 
      (<= (+ j 1) (Length a)))) 
    (invariant 
     (forall m int 
      (==> 
       (&& 
        (<= 0 m) 
        (< m i)) 
       (! 
        (ARRAY_ACCESS a m))))) 
    (invariant 
     (forall n int 
      (==> 
       (&& 
        (< j n) 
        (< n 
         (Length a))) 
       (ARRAY_ACCESS a n)))) 
    (BLOCK 
     (if 
      (! 
       (ARRAY_ACCESS a i)) 
      (BLOCK 
       (:= i 
        (+ i 1))) 
      (BLOCK 
       (if 
        (ARRAY_ACCESS a j) 
        (BLOCK 
         (:= j 
          (- j 1))) 
        (BLOCK 
         (CALL swap 
          (ARGS a i j)) 
         (:= i 
          (+ i 1)) 
         (:= j 
          (- j 1)))))))))))
